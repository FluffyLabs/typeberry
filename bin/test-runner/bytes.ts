function assert(value: boolean, description: string): asserts value is true {
	if (!value) throw new Error(description);
}

export class Blob {
	readonly buffer: ArrayBuffer = new ArrayBuffer(0);
	readonly length?: number = 0;

	constructor(buffer: ArrayBuffer) {
		this.buffer = buffer;
		this.length = buffer.byteLength;
	}

	static parseBlob(v: string): Blob {
		const len = v.length;
		if (len % 2 === 1 || !v.startsWith("0x")) {
			throw new Error(`Invalid hex string: ${v}.`);
		}
		// NOTE [ToDr] alloc
		const buffer = new ArrayBuffer(len / 2 - 1);
		const bytes = new Uint8Array(buffer);
		for (let i = 2; i < len - 1; i += 2) {
			const c = v.substr(i, 2);
			bytes[i / 2 - 1] = Number.parseInt(c, 16);
		}

		return new Blob(buffer);
	}
}

export class Bytes<T extends number> {
	readonly view: DataView = new DataView(new ArrayBuffer(0));
	readonly length: T;

	constructor(view: DataView, len: T) {
		assert(
			view.byteLength === len,
			`Given buffer has incorrect size ${view.byteLength} vs expected ${len}`,
		);
		this.view = view;
		this.length = len;
	}

	static parseBytes<X extends number>(v: string, len: X): Bytes<X> {
		if (v.length > 2 * len + 2) {
			throw new Error(
				`Input string too long. Expected ${len}, got ${v.length / 2 - 1}`,
			);
		}

		const blob = Blob.parseBlob(v);
		return new Bytes(new DataView(blob.buffer), len);
	}
}
