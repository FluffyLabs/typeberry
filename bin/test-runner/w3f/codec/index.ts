import { Header } from "@typeberry/block";
import { type AssurancesExtrinsic, assurancesExtrinsicCodec } from "@typeberry/block/assurances";
import { Block, Extrinsic } from "@typeberry/block/block";
import { DisputesExtrinsic } from "@typeberry/block/disputes";
import { type GuaranteesExtrinsic, guaranteesExtrinsicCodec } from "@typeberry/block/guarantees";
import { type PreimagesExtrinsic, preimagesExtrinsicCodec } from "@typeberry/block/preimage";
import { RefineContext } from "@typeberry/block/refine-context";
import { type TicketsExtrinsic, ticketsExtrinsicCodec } from "@typeberry/block/tickets";
import { WorkReport } from "@typeberry/block/work-report";
import { WorkResult } from "@typeberry/block/work-result";
import { runCodecTest } from "./common";

export async function runAssurancesExtrinsicTest(test: AssurancesExtrinsic, file: string) {
  runCodecTest(assurancesExtrinsicCodec, test, file);
}

export async function runBlockTest(test: Block, file: string) {
  runCodecTest(Block.Codec, test, file);
}

export async function runDisputesExtrinsicTest(test: DisputesExtrinsic, file: string) {
  runCodecTest(DisputesExtrinsic.Codec, test, file);
}

export async function runExtrinsicTest(test: Extrinsic, file: string) {
  runCodecTest(Extrinsic.Codec, test, file);
}

export async function runGuaranteesExtrinsicTest(test: GuaranteesExtrinsic, file: string) {
  runCodecTest(guaranteesExtrinsicCodec, test, file);
}

export async function runHeaderTest(test: Header, file: string) {
  runCodecTest(Header.Codec, test, file);
}

export async function runPreimagesExtrinsicTest(test: PreimagesExtrinsic, file: string) {
  runCodecTest(preimagesExtrinsicCodec, test, file);
}

export async function runRefineContextTest(test: RefineContext, file: string) {
  runCodecTest(RefineContext.Codec, test, file);
}

export async function runTicketsExtrinsicTest(test: TicketsExtrinsic, file: string) {
  runCodecTest(ticketsExtrinsicCodec, test, file);
}

export async function runWorkReportTest(test: WorkReport, file: string) {
  runCodecTest(WorkReport.Codec, test, file);
}

export async function runWorkResultTest(test: WorkResult, file: string) {
  runCodecTest(WorkResult.Codec, test, file);
}
