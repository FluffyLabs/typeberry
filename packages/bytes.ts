function assert(value: boolean, description: string): asserts value is true {
  if (!value) throw new Error(description);
}

function bufferToHexString(buffer: ArrayBuffer): string {
  // TODO [ToDr] consider using TextDecoder API?
  let s = "0x";
  const asUint = new Uint8Array(buffer);
  for (const v of asUint) {
    s += v.toString(16).padStart(2, "0");
  }
  return s;
}

export class BytesBlob {
  readonly buffer: ArrayBuffer = new ArrayBuffer(0);
  readonly length: number = 0;

  constructor(buffer: ArrayBuffer) {
    this.buffer = buffer;
    this.length = buffer.byteLength;
  }

  toString() {
    return bufferToHexString(this.buffer);
  }

  static parseBlob(v: string): BytesBlob {
    const len = v.length;
    if (len % 2 === 1 || !v.startsWith("0x")) {
      throw new Error(`Invalid hex string: ${v}.`);
    }
    // NOTE [ToDr] alloc
    const buffer = new ArrayBuffer(len / 2 - 1);
    const bytes = new Uint8Array(buffer);
    for (let i = 2; i < len - 1; i += 2) {
      const c = v.substring(i, i + 2);
      bytes[i / 2 - 1] = Number.parseInt(c, 16);
    }

    return new BytesBlob(buffer);
  }
}

export class Bytes<T extends number> {
  readonly view: DataView = new DataView(new ArrayBuffer(0));
  readonly length: T;

  constructor(view: DataView, len: T) {
    assert(view.byteLength === len, `Given buffer has incorrect size ${view.byteLength} vs expected ${len}`);
    this.view = view;
    this.length = len;
  }

  toString() {
    return bufferToHexString(this.view.buffer);
  }

  static parseBytes<X extends number>(v: string, len: X): Bytes<X> {
    if (v.length > 2 * len + 2) {
      throw new Error(`Input string too long. Expected ${len}, got ${v.length / 2 - 1}`);
    }

    const blob = BytesBlob.parseBlob(v);
    return new Bytes(new DataView(blob.buffer), len);
  }
}
