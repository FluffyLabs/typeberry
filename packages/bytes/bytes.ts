import { check } from "@typeberry/utils";

function bytesToHexString(buffer: Uint8Array): string {
  // TODO [ToDr] consider using TextDecoder API?
  let s = "0x";
  for (const v of buffer) {
    s += v.toString(16).padStart(2, "0");
  }
  return s;
}

export class BytesBlob {
  readonly buffer: Uint8Array = new Uint8Array([]);
  readonly length: number = 0;

  constructor(data: Uint8Array) {
    this.buffer = data;
    this.length = data.byteLength;
  }

  toString() {
    return bytesToHexString(this.buffer);
  }

  static fromBytes(v: number[]): BytesBlob {
    const arr = new Uint8Array(v);
    return new BytesBlob(arr);
  }

  static parseBlobNoPrefix(v: string): BytesBlob {
    const len = v.length;
    if (len % 2 === 1) {
      throw new Error(`Odd number of nibbles. Invalid hex string: ${v}.`);
    }
    // NOTE [ToDr] alloc
    const buffer = new ArrayBuffer(len / 2);
    const bytes = new Uint8Array(buffer);
    for (let i = 0; i < len - 1; i += 2) {
      const c = v.substring(i, i + 2);
      // TODO [ToDr] [opti] Remove string concat and simply parse each nibble manually
      // (switch from 0..f)
      const parsed = Number(`0x${c}`);
      if (Number.isNaN(parsed)) {
        throw new Error(`Invalid characters in hex byte string: ${c}`);
      }
      bytes[i / 2] = parsed;
    }

    return new BytesBlob(bytes);
  }

  static parseBlob(v: string): BytesBlob {
    if (!v.startsWith("0x")) {
      throw new Error(`Missing 0x prefix: ${v}.`);
    }
    return BytesBlob.parseBlobNoPrefix(v.substring(2));
  }
}

export class Bytes<T extends number> {
  readonly raw: Uint8Array;
  readonly length: T;

  constructor(raw: Uint8Array, len: T) {
    check(raw.byteLength === len, `Given buffer has incorrect size ${raw.byteLength} vs expected ${len}`);
    this.raw = raw;
    this.length = len;
  }

  toString() {
    return bytesToHexString(this.raw);
  }

  isEqualTo(other: Bytes<T>): boolean {
    if (this.length !== other.length) {
      return false;
    }

    for (let i = 0; i < this.length; i++) {
      if (this.raw[i] !== other.raw[i]) {
        return false;
      }
    }
    return true;
  }

  static zero<X extends number>(len: X): Bytes<X> {
    return new Bytes(new Uint8Array(len), len);
  }

  static parseBytesNoPrefix<X extends number>(v: string, len: X): Bytes<X> {
    if (v.length > 2 * len) {
      throw new Error(`Input string too long. Expected ${len}, got ${v.length / 2}`);
    }

    const blob = BytesBlob.parseBlobNoPrefix(v);
    return new Bytes(blob.buffer, len);
  }

  static parseBytes<X extends number>(v: string, len: X): Bytes<X> {
    if (v.length > 2 * len + 2) {
      throw new Error(`Input string too long. Expected ${len}, got ${v.length / 2 - 1}`);
    }

    const blob = BytesBlob.parseBlob(v);
    return new Bytes(blob.buffer, len);
  }
}
