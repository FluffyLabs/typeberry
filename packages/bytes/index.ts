export * from "./bytes";
export * from "./bitvec";
