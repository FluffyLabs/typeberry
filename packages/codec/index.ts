export * from "./decoder";
export * from "./encoder";
export * from "./descriptors";
export { ObjectView, SequenceView } from "./view";

// TODO [ToDr] Temporary re-export bytes until we have proper publishes.
export * as bytes from "@typeberry/bytes";
