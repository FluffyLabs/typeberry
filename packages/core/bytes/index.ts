export * from "./bytes.js";
export * from "./bitvec.js";
