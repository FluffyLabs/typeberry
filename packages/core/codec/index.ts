export * from "./decoder.js";
export * from "./encoder.js";
export * from "./descriptor.js";
export * from "./descriptors.js";
export * from "./validation.js";
export * from "./view.js";
