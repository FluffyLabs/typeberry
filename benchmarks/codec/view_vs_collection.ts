import assert from "node:assert";
import { pathToFileURL } from "node:url";
import { add, complete, configure, cycle, save, suite } from "@typeberry/benchmark/setup.js";
import { Bytes } from "@typeberry/bytes";
import { type CodecRecord, Decoder, type DescribedBy, Encoder, type SequenceView, codec } from "@typeberry/codec";
import { type U64, tryAsU64 } from "@typeberry/numbers";

class TestHeader {
  static Codec = codec.Class(TestHeader, {
    blockNumber: codec.varU64,
    parentHeaderHash: codec.bytes(32),
  });

  static create({ blockNumber, parentHeaderHash }: CodecRecord<TestHeader>) {
    return new TestHeader(blockNumber, parentHeaderHash);
  }

  private constructor(
    public readonly blockNumber: U64,
    public readonly parentHeaderHash: Bytes<32>,
  ) {}
}

const ELEMENTS = 100;
const headerSequence = codec.sequenceVarLen(TestHeader.Codec);
const collection: TestHeader[] = [];
for (let i = 0; i < ELEMENTS; i++) {
  const parentHeaderHash = Bytes.fill(32, i);

  collection.push(TestHeader.create({ blockNumber: tryAsU64(10_000_000n + BigInt(i)), parentHeaderHash }));
}
const encodedData = Encoder.encodeObject(headerSequence, collection);

function compare(
  name: string,
  runDecoded: (s: TestHeader[]) => void,
  runView: (s: SequenceView<TestHeader, DescribedBy<typeof TestHeader.Codec.View>>) => void,
) {
  return [
    add(`Get ${name} from Decoded`, () => {
      const headerSeq = headerSequence.decode(Decoder.fromBytesBlob(encodedData));
      runDecoded(headerSeq);
    }),

    add(`Get ${name} from View`, () => {
      const headerSeq = headerSequence.View.decode(Decoder.fromBytesBlob(encodedData));
      runView(headerSeq);
    }),
  ];
}

export default function run() {
  return suite(
    "Sequence Views",

    ...compare(
      "first element",
      (headerSeq) => {
        assert.deepStrictEqual(headerSeq[0].blockNumber, 10_000_000n);
      },
      (headerSeq) => {
        assert.deepStrictEqual(headerSeq.get(0)?.view().blockNumber.materialize(), 10_000_000n);
      },
    ),

    ...compare(
      "50th element",
      (headerSeq) => {
        assert.deepStrictEqual(headerSeq[50].blockNumber, 10_000_050n);
      },
      (headerSeq) => {
        assert.deepStrictEqual(headerSeq.get(50)?.view().blockNumber.materialize(), 10_000_050n);
      },
    ),

    ...compare(
      "last element",
      (headerSeq) => {
        assert.deepStrictEqual(headerSeq[ELEMENTS - 1].blockNumber, 10_000_000n + BigInt(ELEMENTS - 1));
      },
      (headerSeq) => {
        assert.deepStrictEqual(
          headerSeq
            .get(ELEMENTS - 1)
            ?.view()
            .blockNumber.materialize(),
          10_000_000n + BigInt(ELEMENTS - 1),
        );
      },
    ),

    cycle(),
    complete(),
    configure({}),
    ...save(import.meta.filename),
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run();
}
