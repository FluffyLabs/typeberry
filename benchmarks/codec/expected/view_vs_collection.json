{
  "name": "Sequence Views",
  "date": "2024-12-08T09:35:35.899Z",
  "version": null,
  "results": [
    {
      "name": "Get first element from Decoded",
      "ops": 31173.02,
      "margin": 0.56
    },
    {
      "name": "Get first element from View",
      "ops": 68299.98,
      "margin": 0.2
    },
    {
      "name": "Get 50th element from Decoded",
      "ops": 31269.64,
      "margin": 0.45
    },
    {
      "name": "Get 50th element from View",
      "ops": 38404.07,
      "margin": 2.02
    },
    {
      "name": "Get last element from Decoded",
      "ops": 31315.47,
      "margin": 0.33
    },
    {
      "name": "Get last element from View",
      "ops": 27069.69,
      "margin": 0.6
    }
  ],
  "fastest": {
    "name": "Get first element from View",
    "index": 1
  },
  "slowest": {
    "name": "Get last element from Decoded",
    "index": 4
  }
}
